src/github/task-workflow.ts

```diff
@@ -0,0 +1,220 @@
+
+  /**
+   * Steps to run before the main build step.
+   *
+   * @default - not set
+   */
+  readonly preBuildSteps?: JobStep[];
+
+  /**
+   * The main task to be executed.
+   */
+  readonly task: Task;
+
+  /**
+   * Main build step used in the workflow.
+   *
+   * @default - by default we will run `projen ${task.name}`
+   */
+  readonly buildStep?: JobStep;
+
+  /**
+   * Actions to run after the main build step.
+   *
+   * @default - not set
+   */
+  readonly postBuildSteps?: JobStep[];
+
+  /**
+   * Enables anti-tamper checks in the workflow.
+   *
+   * @default true
+   */
+  readonly antitamper?: boolean;
+
+  /**
+   * Actions to run as the last step in the job.
+   *
+   * @default - not set
+   */
+  readonly finalSteps?: JobStep[];
+
+  /**
+   * Workflow environment variables.
+   * @default {}
+   */
+  readonly env?: Record<string, string>;
+
+  /**
+   * Permissions for the build job.
+   */
+  readonly permissions: JobPermissions;
+}
+
+/**
+ * A GitHub workflow for common build tasks within a project.
+ */
+export class TaskGithubWorkflow extends GithubWorkflow {
+  readonly github: GitHub;
+  readonly jobId: string;
+
+  constructor(github: GitHub, options: TaskGithubWorkflowOptions) {
+    super(github, options.name);
+    this.jobId = options.jobId ?? DEFAULT_JOB_ID;
+    this.github = github;
+    this.createWorkflow(options);
+  }
+
+  protected getMainStep(options: TaskGithubWorkflowOptions) {
+    return options.buildStep ?? {
+      name: options.task.name,
+      run: this.github.project.runTaskCommand(options.task),
+    };
+  }
+
+  protected createWorkflow(options: TaskGithubWorkflowOptions) {
+    if (options.trigger) {
+      if (options.trigger.issueComment) {
+        // https://docs.github.com/en/actions/learn-github-actions/security-hardening-for-github-actions#potential-impact-of-a-compromised-runner
+        throw new Error('Trigger "issueComment" should not be used due to a security concern');
+      }
+
+      this.on(options.trigger);
+    }
+
+    this.on({
+      workflowDispatch: {}, // allow manual triggering
+    });
+
+    const preCheckoutSteps = options.preCheckoutSteps ?? [];
+    const checkoutWith = options.checkoutWith ? { with: options.checkoutWith } : {};
+    const preBuildSteps = options.preBuildSteps ?? [];
+    const postBuildSteps = options.postBuildSteps ?? [];
+    const finalSteps = options.finalSteps ?? [];
+
+    const antitamper = options.antitamper ?? true;
+    const antitamperSteps = antitamper ? [{
+      name: 'Anti-tamper check',
+      run: 'git diff --ignore-space-at-eol --exit-code',
+    }] : [];
+
+    if (options.artifactsDirectory) {
+      finalSteps.push({
+        name: 'Upload artifact',
+        uses: 'actions/upload-artifact@v2.1.1',
+        if: 'always()',
+        with: {
+          name: options.artifactsDirectory,
+          path: options.artifactsDirectory,
+        },
+      });
+    }
+
+    const job: Mutable<Job> = {
```

Comment:
Does it have to be mutable?

```suggestion
    const job: Job = {
```

Reply:
it does not